src/python/pants/engine/internals/build_files.py

```diff
@@ -387,16 +389,88 @@ def _extract_env_vars(
     )
 
     # Process synthetic targets.
-    for address_map in address_maps:
-        for synthetic in synthetic_address_maps:
-            synthetic.process_declared_targets(address_map)
-            synthetic.apply_defaults(frozen_defaults)
+
+    def apply_defaults(tgt: TargetAdaptor) -> TargetAdaptor:
+        default_values = frozen_defaults.get(tgt.type_alias)
+        if default_values is None:
+            return tgt
+        return tgt.with_new_kwargs(**default_values, **tgt.kwargs)
+
+    name_to_path_and_synthetic_target: dict[str, tuple[str, TargetAdaptor]] = {}
```

Comment:
This new block is modified from the now-deleted `SyntheticAddressMap.process_declared_targets`. 

This requires some careful eyeballing, ideally by @kaos, 
who authored the original logic.

Having this logic as a mutator on `SyntheticAddressMap` was
broken for two reasons:

1) That class is intended to be hashable and therefore immutable.
2) That method mutated not just `self`, but the `address_map` argument.  But we applied it in a loop over multiple AddressMap instances, so the behavior was order-dependent. The only reason we never noticed this bug is that we don't use `_extend_synthetic` in practice, and we did not adequately test for that case.

(I plan to add a test that covers the issue)

Reply:
Update: test added